source/blog/2024-11-05-hanami-220.html.markdown

```diff
@@ -0,0 +1,244 @@
+
+- Share a single database, but have each slice provide its own relations, so they can choose exactly how much of the database to expose. This is our default.
+- Use a dedicated database for certain slices. This is as easy as a `SLICE_NAME__DATABASE_URL` env var!
+- Connect to any number of _additional_ databases within your app or slice. This is as easy as creating a `SLICE_NAME__DATABASE_URL__GATEWAY_NAME` env var!
+- Adopt a simpler, blended development experience by sharing a single database and set of relations across all slices.
+- Or any combination of the above!
+
+## Smooth operations
+
+Along with Hanami 2.2, we’re excited to debut the brand new [dry-operation](https://dry-rb.org/gems/dry-operation) gem. With dry-operation, you have a streamlined tool for organizing your business logic into flexible, composable objects made from flows of internal steps.
+
+Creating an operation is as easy as `hanami generate operation posts.create_post`. Operations can be built from multiple steps, with each returning a `Result`:
+
+```ruby
+module MyApp
+  module Posts
+    class CreatePost < MyApp::Operation
+      include Deps["repos.post_repo"]
+
+      def call(attributes)
+        validation = step validate(attributes)
+
+        post = post_repo.create(validation.to_h)
+
+        Success(post)
+      end
+
+      private
+
+      def validate(attributes)
+        # Validate attributes here.
+
+        # Return a `Failure` and execution above will short-circuit
+        #   Failure(errors: ["not valid"])
+
+        # Return a `Success` and execution will continue with the value unwrapped
+        #   Success(attributes)
+      end
+    end
+  end
+end
+```
+
+Every operation returning a `Success` or `Failure` is great for consistency (every caller is required to consider both sides), but also for expressiveness. You can now turn to pattern matching on results in your actions, for example:
+
+```ruby
+module MyApp
+  module Actions
+    module Posts
+      class Create < MyApp::Action
+        include Deps["posts.create_post"]
+
+        def handle(request, response)
+          result = create_post.call(request.params[:post])
+
+          case result
+          in Success(post)
+            response.redirect_to routes.path(:post, post.id)
+          in Failure(validation)
+            response.render(view, validation:)
+          end
+        end
+      end
+    end
+  end
+end
+```
+
+Operations are natively integrated with Hanami’s database layer, providing `transaction do ... end` blocks to ensure database changes are written together, with any intervening `Failure` automatically rolling back the transaction.
+
+dry-operation is the long-awaited successor to the venerable [dry-transaction](http://dry-rb.org/gems/dry-transaction) gem, and I’m deeply grateful to [Marc Busqué](https://github.com/waiting-for-dev) for building it.
+
+## More up our sleeve
```

Comment:
Should we add something about supporting Ruby 3.1+?

Should we add a link to our Discourse, and maybe a reference to their being a chat on Discourse that poeple can see only once they create an account?

Reply:
Good point, added "Dropped support for Ruby 3.0"